### PR TITLE
Add support for Windows (Fixes #28)

### LINKED ISSUE #28: Windows Support
Support Windows via the Linux subsystem

### PR DESCRIPTION
**Pull request by zacps**
Primarily this:
* Adds Windows support to copypasta
* Adds rusttype as a font rendering backend
* Creates an interface to abstract over how we interact with a TTY-like thing (EventedRW)
* Adds safe winpty bindings
* Adds [winpty](https://github.com/rprichard/winpty) as a TTY shim
* Adds appveyor CI support

The windows support is not perfect yet but it's good enough that I think it makes sense to merge it. Any unresolved issues are listed [here](https://github.com/zacps/alacritty/issues).

Apologies for the unrelated formatting changes, I can dig through and remove them from the pull but it might take me a little while.

### CODE REVIEW

**Comment by chrisduerr** [comment]
Thanks a ton for working on this! Great to see a PR being opened for this. :)

It will probably take a while until this is merged because I'm pretty sure the focus is on scrollback right now, so feel free to keep working on the remaining issues in this PR. But it looks like none of these should be blockers? However any issue remains before this is merged, the issues should be transferred to this repository.

File: README.md [hunk 1]
```diff
@@ -207,18 +205,6 @@ It might be handy to mask all other packages provided in the `slyfox` overlay by
...
-If you have a rust toolchain setup you can install Alacritty via cargo:
-
-```sh
-cargo install --git https://github.com/jwilm/alacritty
-```
-
-Note that you still need to download system build dependencies via your package
-manager as mentioned above. The binary `alacritty` will be placed into `$HOME/.cargo/bin`.
-Make sure it is in your path (default if you use `rustup`).
-
```
Comment (chrisduerr) [inline README.md]: What's the reason for removing this?

File: README.md [hunk 2]
```diff
@@ -341,35 +329,25 @@ Just Works.
...
-  Alacritty's original purpose was to provide a better experience when using
-  [tmux] which already handled scrollback. The scope of this project has since
-  expanded, and [scrollback will eventually be added](https://github.com/jwilm/alacritty/issues/124).
+- _My arrow keys don't work_. It sounds like you deleted some key bindings from
+  your config file. Please reference the default config file to restore them.
+- _Why doesn't it support scrollback?_ Alacritty's original purpose was to
+  provide a better experience when using [tmux] which already handled
+  scrollback. The scope of this project has since expanded, and [scrollback will
+  eventually be added](https://github.com/jwilm/alacritty/issues/124).
```
Comment (chrisduerr) [inline README.md]: I don't think these changes are necessary?

File: README.md [hunk 3]
```diff
@@ -238,8 +224,11 @@ cargo build --release
 ```
 
 If all goes well, this should place a binary at `target/release/alacritty`.
+**BEFORE YOU RUN IT:** Install the config file as described below; otherwise,
+many things (such as arrow keys) will not work. If you're on macOS, you'll need
+to change the `monospace` font family to something like `Menlo`.
```
Comment (chrisduerr) [inline README.md]: I believe this is also mentioned in a different place of the config, potentially there has been a merge issue?

File: copypasta/src/windows.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+impl ::std::fmt::Display for Error {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        match *self {
+            Error::Clipboard(ref err) => err.fmt(f),
+        }
+    }
+}
+
+// TODO: Is this really safe?
```
Comment (chrisduerr) [inline copypasta/src/windows.rs]: This should probably be properly evaluated before merging. I haven't taken a close at it myself yet, just leaving a comment for tracking here.

File: src/input.rs [hunk 5]
```diff
@@ -206,15 +207,24 @@ impl Action {
...
                         unsafe { ::libc::daemon(1, 0); }
                         Ok(())
                     })
-                    .spawn()
+                    .spawn();
+
+                #[cfg(windows)]
+                let spawned = Command::new(program)
+                    .args(args)
+                    .spawn();
```
Comment (chrisduerr) [inline src/input.rs]: Instead of having different implementations, maybe the better choice is to properly `wait` for the process exit? That should work on both unix and windows.

Comment (zacps) [reply src/input.rs]: I'm not actually sure of the purpose of having a command bound to a key, and haven't tested it myself.

**Comment by zacps** [comment]
Yeah they were mostly merge issues, I haven't quite mastered git yet :wink:.

I forgot to mention in the original post but it'd be good if someone could take a look at the rust build script for [winpty](https://github.com/zacps/winpty/tree/rust) because I don't want to create a pull request for a foreign language without it getting some review first.

**Comment by jwilm** [comment]
@zacps this is incredibly exciting for me to see, and I know there is a good number of people wanting to use Alacritty on windows but just waiting for support to land. Thank you very much for not only working on this but actually delivering a fully functional PR!

I'm looking forward to taking this for a spin and adding a review. My inclination here is to pull this in without too much delay since it's largely _new_ work that won't impact existing platforms. Only caveat there is that I won't be able to make time to do so until June 29 at the earliest.

On a quick glance through the files, I noticed a dll and a couple of exes were added to the tree. Is it possible we could provide these through a build step instead of adding to the git tree?

One final note, I would very much like to invite you to join the project as a collaborator as our resident Windows expert, if you are open to it.

Thanks again!

**Comment by zacps** [comment]
So the dll is just a 32bit libclang required for 32 bit appveyor builds (bindgen dep). I'll try change the appveyor config to pull it down during the build.

The exe is unfortunately a bit different. Winpty is both an exe (that does the actual work) and a library to communicate with the exe. The library and bindings are included in the build step. The problem is that cargo has no way to 'publish' artifacts it doesn't understand. So while we can build the exe in the dependency there's no way to make it accessible to alacritty. The built file will be located in one of the target directories 'somewhere'.

Now that I think about it we could traverse the target directory and look for the exe with the latest timestamp, I'm not sure if that'd be worth doing.

I think I'll create an issue on the cargo repo so they're aware of this use case.

I'd love to join as a collaborator :smiley:

**Comment by binary132** [comment]
@zacps @jwilm, perhaps those binary dependencies should be packaged via the make step?

Storing an executable dependency with a binary shared object in Git strikes me as a particularly egregious antipattern.

**Comment by zacps** [comment]
binary132 It's definitely not ideal.

As I said in the previous comment the main issue is that cargo doesn't have any way to 'push' an artifact it doesn't understand up the dependency tree.

```
winpty-sys <- Binary dep built here 
    winpty 
        alacritty <- Binary needed here 
```

Alacritty doesn't know where the built artifacts ended up. We can potentially try to search from them in target but that depends on the structure of target being stable (I don't know if it is).

The ideal solution is for a build script to be able to indicate that it wants to move an artifact into it's parent's target directory.

**Comment by zacps** [comment]
I also played around trying to implement finding the Binary and I realized we'd have to add [gyp](https://gyp.gsrc.io/) as another manually installed dependency.

**Comment by Boddlnagg** [comment]
You probably have already heard about the new [ConPTY](https://blogs.msdn.microsoft.com/commandline/2018/08/02/windows-command-line-introducing-the-windows-pseudo-console-conpty/), but I'm adding a note here anyway. ConPTY is probably a saner way to implement this on Windows, but it will only be supported starting from the version of Windows 10 that's released in winter/fall 2018.

**Comment by prabirshrestha** [comment]
Would be good if we could have just one config file instead of `alacritty_windows.yml`. Then all I would need is the following config in my `alacritty.yml` file.

```
os:
    windows:
         shell:
              program: cmd
    
```

**Comment by jonathandturner** [comment]
Anywhere folks can jump in to help make this happen?

**Comment by zacps** [comment]
I think we're just waiting on review from @jwilm at this point.

**Comment by jwilm** [comment]
This is next on my list to review. One question I have is whether there's a way we can provide winpty.exe without committing a big binary to git, perhaps through a build process. Would love some ideas/feedback on this. I know @zacps has already spent some time investigating, but maybe you have some other ideas @jonathandturner? There's some discussion above.

**Comment by jonathandturner** [comment]
I don't, sorry. Short of, say, fetching it with a build.rs or somesuch.

**Comment by zacps** [comment]
Would you prefer adding gyp as a build dependency and doing some path-hacking over including prebuilt binaries? Those are the only two options as far as I can tell.

**Comment by jwilm** [comment]
@zacps I had another thought that we could maybe leverage git-lfs. Only problem is that GitHub charges an insane amount for that, and I would want to look into a self-hosted solution first.

For the gyp path, how do you estimate the complexity of that? If we don't have gyp, how hard would it be to regenerate these artifacts when the upstream changes?

**Comment by jwilm** [comment]
Er, maybe git-lfs isn't that badly priced. I can't tell for certain, but it looks like $5 for 50 GB, which isn't too horrible.. might want to start shipping binaries before going down that path, though.

**Comment by dpastoor** [comment]
Don’t forget gh also charges for bandwidth and forks don’t have seamless
access I don’t believe. IMO gyp and things of that nature are stable enough
to be non versioned and put in like an s3 bucket or the like

**Comment by prabirshrestha** [comment]
Could download the github file from download release or from nuget server which is npm for .NET. https://www.nuget.org/packages/winpty.NET/

Here is a download link: https://www.nuget.org/api/v2/package/winpty.NET/0.4.2 
nupkg is just a zip file so should be easy to extract it. 

Would still be good to check the hash before extracting.

**Comment by zacps** [comment]
@prabirshrestha Huh I didn't know that existed, it includes the agent executable as well.

I think it probably makes sense to include this in Alacritty's build script rather than the winpty bindings.

I'll add that and try rebase it today/tomorrow.

**Comment by fabienheureux** [comment]
Hi @zacps, really excited to see this work on Alacritty for Windows ! 

Are you using Windows Subsystem for Linux inside Alacritty ? 
How could I build and try the Alacritty's version from your PR ? I read the README you updated, but did not notice any instructions, nor Makefile changes windows-specific (and I am not very experienced with Windows development). 

Thanks a lot for your work !

**Comment by zacps** [comment]
@jwilm Do you have any ETA on when you can take a look at this?
It would be great to get this reviewed soon, following master is not no work.

**Comment by joelhandwell** [comment]
@zacps while waiting for review, do you think it's worthy to provide chocolatey package for alacritty ?
Hopefully install it with ```choco install alacritty``` command.

**Comment by zacps** [comment]
@joelhandwell It's not something I'll do myself, but you're welcome to if you want.

**Comment by jwilm** [comment]
@zacps I'm starting to run this branch locally on Linux just to check for any regressions. One thing I noticed initially is that scrolling-in-region bench seems to be slightly slower than master.

I'll try and do a review tomorrow morning after playing with it today. Maybe @chrisduerr can also take a look at the code. Note that I'm going to attempt to avoid a lot of nit-picking in the interests of getting this in-tree sooner than later.

Thanks for making the changes to move the binary artifacts out of git.

File: .travis.yml [hunk 6]
```diff
@@ -18,6 +18,11 @@ env:
 install:
   - if [ -n "$CLIPPY" ]; then rustup component add clippy-preview; fi
 
+addons:
+  apt:
+    packages:
+      gperf
+
```
Comment (chrisduerr) [inline .travis.yml]: This was necessary for a while, but I think it should be good even without it now.

Current master seems to build fine without it.

File: build.rs [hunk 7]
```diff
@@ -11,20 +11,75 @@
...
+    let target = match env::var("TARGET").unwrap().split("-").next().unwrap() {
+        "x86_64" => "x86",
+        "i386" => "x64",
+        _ => panic!("architecture has no winpty binary")
+    };
+
+    let mut winpty_agent = archive.by_name(&format!("content/winpty/{}/winpty-agent.exe", target)).unwrap();
+
+    io::copy(&mut winpty_agent, &mut File::create(out_path).unwrap()).unwrap();
+}
```
Comment (chrisduerr) [inline build.rs]: Missing newlines at the end of file seem to be a common issue.

It's a bit of a small nitpick, but it should be easy enough to resolve.

File: src/tty/mod.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+#[cfg(windows)]
+mod windows;
+#[cfg(windows)]
+pub use self::windows::*;
+
+/// This trait defines the behaviour needed to read and/or write to a stream.
+/// It defines an abstraction over mio's interface in order to allow either one
+/// read/write object or a seperate read and write object.
+pub trait EventedRW {
```
Comment (chrisduerr) [inline src/tty/mod.rs]: The name for this trait is a bit odd. Maybe `Event` would be better than `Evented` here? I'm not 100% what exactly is meant by 'evented'. Also `ReadWrite` is short enough to write it out completely imho. I think `EventReadWrite` would be a bit closer to Rust's guidelines for trait names.

**Comment by jwilm** [comment]
Regarding perf, I was testing with latest changes from master. Whatever the difference is with scrolling performance, it is real.

File: src/tty/mod.rs (hunk 8 above)
Comment (zacps) [reply src/tty/mod.rs]: `Evented` refers to the fact that it implements `mio::event::Evented`, so I think it makes sense to maintain that?

Changing `RW` to `ReadWrite` makes sense though.

**Comment by zacps** [comment]
Are you using vtebench to test performance? I'll give it a look.

File: appveyor.yml [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+install:
+  - curl -sSf -o rustup-init.exe https://win.rustup.rs/
+  - rustup-init.exe -y --default-host %TARGET% --default-toolchain %RUST_VERSION%
+  - set PATH=%PATH%;C:\Users\appveyor\.cargo\bin
+  - rustc -Vv
+  - cargo -V
+  # Make 32bit libclang dll available for winpty-sys build on i686
+  - If [%TARGET%]==[i686-pc-windows-msvc] (
+      copy assets\windows\x86\libclang_win32_5.0.1.dll "C:\Program Files\LLVM\bin\libclang.dll"
```
Comment (jwilm) [inline appveyor.yml]: Please remove the libclang.dll from git as well.

File: src/tty/mod.rs (hunk 8 above)
Comment (chrisduerr) [reply src/tty/mod.rs]: Oh okay, I wasn't aware that `Evented` is a trait from `mio`. Should be fine to use `EventedReadWrite`.

File: build.rs (hunk 7 above)
Comment (chrisduerr) [reply build.rs]: You've resolved it for this file, however it's still present in a lot of other files. Github's little icon should help showing where exactly it's required.

**Comment by zacps** [comment]
Testing with:

```sh
vtebench -w $(tput cols) -h $(tput lines) -b=100000000 scrolling-in-region  > scroll_test.vte
```

On `windows-linux`:

```sh
$ time cat scroll_test.vte
cat ./scroll_test.vte  0.00s user 7.41s system 99% cpu 7.476 total
```

On `master`:

```sh
$ time cat scroll_test.vte
cat scroll_test.vte  0.00s user 7.44s system 98% cpu 7.533 total
```

Could you tell me what settings you're using to test? It seems comparable on my machine.

**Comment by jcdickinson** [comment]
One quick addition, if you look at the the build script for winpty.Net (the nuget package being used for the winpty binaries) it is doing the following:

```bat
curl -#Lo winpty-native.zip "https://github.com/rprichard/winpty/releases/download/0.4.2/winpty-0.4.2-msvc2015.zip"
call 7z e -y -o"content\winpty\x86" winpty-native.zip "ia32\bin\winpty.dll" > nul
call 7z e -y -o"content\winpty\x86" winpty-native.zip "ia32\bin\winpty-agent.exe" > nul
call 7z e -y -o"content\winpty\x86" winpty-native.zip "ia32\bin\winpty-debugserver.exe" > nul
call 7z e -y -o"content\winpty\x64" winpty-native.zip "x64\bin\winpty.dll" > nul
call 7z e -y -o"content\winpty\x64" winpty-native.zip "x64\bin\winpty-agent.exe" > nul
call 7z e -y -o"content\winpty\x64" winpty-native.zip "x64\bin\winpty-debugserver.exe" > nul
```

If you want, you can skip nuget entirely if you just grab the binary from Github (0.4.3 is current as of this comment).

**Comment by zacps** [comment]
I don't think it's any better really. We still have to download a zip then extract a particular file.

**Comment by jwilm** [comment]
@zacps one additional argument

```
vtebench -w $(tput cols) -h $(tput lines) -b=100000000 scrolling-in-region --lines-from-bottom=2 > scroll_test.vte
```

**Comment by jwilm** [comment]
The weird thing to me is that I don't see anything that should obviously cause a difference.

**Comment by jwilm** [comment]
@zacps possibly; what was the motivation for that change?

**Comment by zacps** [comment]
Nevermind, I realised it's not that because it doesn't affect all benchmarks.
On that note I was not able to reproduce at all.

Given the test generated by @jwilm 's command above I got a difference of less than 100ms (with a total of about 15.7secs).

**Comment by zacps** [comment]
Using jwilm's command again:
image

Does that look ok to you? It's going to be hard for me to narrow down the cause when it's this small, if there even is a difference.

**Comment by jwilm** [comment]
@zacps what drove the change of edge to level triggered in the event loop?

File: font/src/lib.rs [hunk 10]
```diff
@@ -336,12 +342,16 @@ pub trait Rasterize {
...
+    /// Create a new Rasterizer
     fn new(device_pixel_ratio: f32, use_thin_strokes: bool) -> Result<Self, Self::Err>
-        where Self: Sized;
+    where
+        Self: Sized;
 
     /// Get `Metrics` for the given `FontKey`
+    #[cfg(not(windows))]
     fn metrics(&self, FontKey) -> Result<Metrics, Self::Err>;
+    #[cfg(windows)]
```
Comment (jwilm) [inline font/src/lib.rs]: If `Size` is needed in one of these APIs, we should just pass it in both.

File: src/config.rs [hunk 11]
```diff
@@ -1472,6 +1475,19 @@ impl Config {
             .map(|path| path.into())
     }
 
+    #[cfg(windows)]
```
Comment (jwilm) [inline src/config.rs]: We shouldn't be conditionally compiling this entire method; XDG should be the default on Windows as well. Perhaps the final fallback path needs to be computed per-platform, though.

File: src/config.rs [hunk 12]
```diff
@@ -1480,6 +1496,15 @@ impl Config {
         Ok(path.into())
     }
 
+    #[cfg(windows)]
```
Comment (jwilm) [inline src/config.rs]: This should be able to go away when using XDG paths, right?

File: src/event_loop.rs [hunk 13]
```diff
@@ -323,56 +313,53 @@ impl<Io> EventLoop<Io>
...
-        state: Option<State>
-    ) -> thread::JoinHandle<(EventLoop<Io>, State)> {
+    pub fn spawn(mut self, state: Option<State>) -> thread::JoinHandle<(Self, State)> {
         thread::spawn_named("pty reader", move || {
             let mut state = state.unwrap_or_else(Default::default);
             let mut buf = [0u8; 0x1000];
 
-            let fd = self.pty.as_raw_fd();
-            let fd = EventedFd(&fd);
+            let poll_opts = PollOpt::level() | PollOpt::oneshot();
```
Comment (jwilm) [inline src/event_loop.rs]: There's edge triggering used elsewhere in the code still. Why was this changed to level?

File: src/input.rs [hunk 14]
```diff
@@ -232,15 +233,24 @@ impl Action {
             },
             Action::Command(ref program, ref args) => {
                 trace!("running command: {} {:?}", program, args);
-                match Command::new(program)
+
+                #[cfg(not(windows))]
+                let spawned = Command::new(program)
                     .args(args)
                     .before_exec(|| {
```
Comment (jwilm) [inline src/input.rs]: Please `cfg` the `before_exec` method rather than the entire `Command::new` call.

File: src/tty/windows.rs [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+
+use tty::EventedReadWrite;
+use term::SizeInfo;
+use display::OnResize;
+use config::{Config, Shell};
+use cli::Options;
+use mio;
+use std::io;
+
+use std::os::raw::c_void;
```
Comment (jwilm) [inline src/tty/windows.rs]: Please clean up imports in this file.

```
// std imports

// crate imports

// local imports
```

File: src/lib.rs [hunk 16]
```diff
@@ -50,6 +61,8 @@ extern crate unicode_width;
 extern crate vte;
 extern crate xdg;
 extern crate base64;
+extern crate winit;
+extern crate image;
```
Comment (jwilm) [inline src/lib.rs]: Can this be conditionally imported as well? It looks like it's only used on Windows.

File: src/renderer/mod.rs [hunk 17]
```diff
@@ -174,16 +170,20 @@ impl GlyphCache {
...
+    where
+        L: LoadGlyph,
     {
         let (regular, bold, italic) = Self::compute_font_keys(font, &mut rasterizer)?;
 
         // Need to load at least one glyph for the face before calling metrics.
         // The glyph requested here ('m' at the time of writing) has no special
         // meaning.
         rasterizer.get_glyph(GlyphKey { font_key: regular, c: 'm', size: font.size() })?;
+        #[cfg(windows)]
```
Comment (jwilm) [inline src/renderer/mod.rs]: Can remove after addressing other changes.

File: src/renderer/mod.rs [hunk 18]
```diff
@@ -267,6 +264,14 @@ impl GlyphCache {
         FontDesc::new(&desc.family[..], style)
     }
 
+    #[cfg(windows)]
```
Comment (jwilm) [inline src/renderer/mod.rs]: Can remove after addressing other changes.

File: src/renderer/mod.rs [hunk 19]
```diff
@@ -289,8 +294,8 @@ impl GlyphCache {
                 rasterized.top += i32::from(glyph_offset.y);
                 rasterized.top -= metrics.descent as i32;
 
-                loader.load_glyph(&rasterized)
-            })
+            loader.load_glyph(&rasterized)
```
Comment (jwilm) [inline src/renderer/mod.rs]: This indentation change seems wrong

File: src/config.rs (hunk 11 above)
Comment (zacps) [reply src/config.rs]: The [XDG](https://github.com/whitequark/rust-xdg) library does not support windows which makes sense, because it's only applicable for UNIX.

I did find [directories-rs](https://github.com/soc/directories-rs) which has specific windows support.

File: src/input.rs (hunk 14 above)
Comment (zacps) [reply src/input.rs]: As far as I can tell you can't? Attributes can be applied to expressions, but `.before_exec(...)` isn't an expression.

Doing that fails for me:

```rust
error: expected one of `.`, `;`, `?`, or an operator, found `#`
   --> src\input.rs:239:21
    |
238 |                     .args(args)
    |                                - expected one of `.`, `;`, `?`, or an operator here
239 |                     #[cfg(not(windows))]
    |                     ^ unexpected token
```

Comment (zacps) [reply src/input.rs]: You can't put an attribute on `.before_exec(...)` because it isn't an expression, and you can't put the attribute on the lib call because `before_exec` is a UNIX only extension.

How were you thinking of doing it?

File: src/event_loop.rs (hunk 13 above)
Comment (zacps) [reply src/event_loop.rs]: There's no good reason, it was just what I changed that happened to make it work. I've tried a number of times to figure it out but haven't been able to.

Basically something about the change (5b444375 -> windows-linux) meant that we don't read all of the data from the TTY Read pipe/input channel and block on the next `poll.poll` call.

File: src/config.rs (hunk 11 above)
Comment (jwilm) [reply src/config.rs]: Interesting; maybe we should have a follow-up issue to switch over.

File: src/input.rs (hunk 14 above)
Comment (jwilm) [reply src/input.rs]: Ahh, I didn't realize `before_exec` was only on UNIX systems. I figured we could have defined functions instead of a closure and use an empty one on Windows.

Since that's not an option, this is fine for now.

File: src/event_loop.rs (hunk 13 above)
Comment (jwilm) [reply src/event_loop.rs]: Hmm, no idea what that could have been. Before we merge this, I want to try the benchmarks with edge triggering on linux again just to see if it has an impact. There's nothing else in this PR that seems like it should have any effect on performance.

File: src/config.rs (hunk 11 above)
Comment (zacps) [reply src/config.rs]: I've created an [issue](https://github.com/zacps/alacritty/issues/27) on my fork which will be copied over when we merge.

File: src/event_loop.rs (hunk 13 above)
Comment (zacps) [reply src/event_loop.rs]: Fixed in 5dcd5cd. It was actually an issue with not reregistering the input channel which then didn't recieve any more events because of `PollOpt::Oneshot`.

Haven't tested performance, my guess is it won't change.

Comment (zacps) [reply src/event_loop.rs]:
```shell
❯ hyperfine "./alacritty_master -e '/usr/bin/cat' ./scroll_test.vte" "./alacritty_windows -e '/usr/bin/cat' ./scroll_test.vte"
Benchmark #1: ./alacritty_master -e '/usr/bin/cat' ./scroll_test.vte
  Time (mean ± σ):      7.811 s ±  0.157 s    [User: 2.727 s, System: 11.529 s]
  Range (min … max):    7.647 s …  8.090 s
Benchmark #2: ./alacritty_windows -e '/usr/bin/cat' ./scroll_test.vte
  Time (mean ± σ):      7.915 s ±  0.189 s    [User: 2.761 s, System: 11.609 s]
  Range (min … max):    7.660 s …  8.274 s
Summary
  './alacritty_master -e '/usr/bin/cat' ./scroll_test.vte' ran
    1.01 ± 0.03 times faster than './alacritty_windows -e '/usr/bin/cat' ./scroll_test.vte'
```

Comment (zacps) [reply src/event_loop.rs]: From this I'm going to say the performance is equal. 

I even bumped the number of runs from 10 to 20 and the difference range increased (was `± 0.03`).

```shell
❯ hyperfine "./alacritty_master -e '/usr/bin/cat' ./scroll_test.vte" "./alacritty_windows -e '/usr/bin/cat' ./scroll_test.vte" --min-r
Benchmark #1: ./alacritty_master -e '/usr/bin/cat' ./scroll_test.vte
  Time (mean ± σ):      7.921 s ±  0.264 s    [User: 2.791 s, System: 11.651 s]
  Range (min … max):    7.585 s …  8.532 s
Benchmark #2: ./alacritty_windows -e '/usr/bin/cat' ./scroll_test.vte
  Time (mean ± σ):      7.842 s ±  0.208 s    [User: 2.706 s, System: 11.577 s]
  Range (min … max):    7.606 s …  8.396 s
Summary
  './alacritty_windows -e '/usr/bin/cat' ./scroll_test.vte' ran
    1.01 ± 0.04 times faster than './alacritty_master -e '/usr/bin/cat' ./scroll_test.vte'

```

**Comment by zacps** [comment]
@jwilm I think I've addressed all the issues you pointed out.

**Comment by jwilm** [comment]
@zacps thanks! I'll run with a build off this branch until I get an opportunity to bench/final review.

@chrisduerr let's try and avoid landing any major changes until we get this in-tree.

**Comment by chrisduerr** [comment]
I'm fine with not landing any major changes, however the performance issue is still present.

I'd like to make sure the remaining performance gaps are closed, since there isn't really any good reason why this should impact performance on linux/macos (those users aren't getting any benefit from losing that performance).

In my benchmarks, I've been able to see around 5% performance degradation in the `scrolling-in-region` benchmark. I wasn't able to see anything in the other benchmarks, however I've just compared single data points instead of a range of results.

**Comment by zacps** [comment]
@chrisduerr Could you post some information about your system?

My system:

```
OS: Arch linux
CPU: i5 4690k @4.5GHz
GPU: GTX970 (Proprietary drivers)
```

Also please try running multiple benchmarks, using the command I ran above with `hyperfind` makes this a lot easier. Even across 20 runs of ~70sec long tests my branch had a smaller *minimum* time than *master*. One run will not give enough information to draw a conclusion.

**Comment by chrisduerr** [comment]
I'm running a i7 6700k system with the mesa driver (R9 390). Note that I did not run just a single benchmark. The performance degradation is significant and reproducible. And based on what jwilm has said, it doesn't seem like it's something specifically with my system either.

Note that for reproducing this, I would recommend setting a relatively big distance from bottom for the scroll in region benchmark. With my tests that was able to showcase the regression best.

**Comment by zacps** [comment]
I still can't reproduce, I bumped it up to `5` and `15` lines from bottom and if anything the windows branch performed slightly better (though not a significant difference).

**Comment by chrisduerr** [comment]
Okay, so I've gone over this and tested things again and again. And I think I'd recommend to go forward with this PR without further investigations of the performance issues.

I've tested running standalone benchmarks, hyperfine, running it through `perf` and using the benchmarks created in #1653 and I was not able to reliably reproduce any performance degradations.

When running standalone benchmarks there were some interesting fluctuations, like a slowdown of .2s when running the scrolling-in-region benchmark with an offset of 50 after running the asrw benchmark. However I wasn't able to reproduce this consistently and I think that can be mostly blamed on CPU frequency adjustments.

Unless @jwilm is able to reliably reproduce a performance regression, I don't think it's worth to put any more effort into this. Especially because the offset in a scrolling region usually isn't bigger than 1 or 2, so performance regressions with an offset of >10 shouldn't be incredibly critical anyways.

**Comment by jwilm** [comment]
@zacps thanks for all the effort here. I invited you as a collaborator to the project so that you're empowered to improve Windows support and manage issues/PRs pertaining to it.

**Comment by jwilm** [comment]
@zacps Feel free to move the issues over from your tracker and label them with `H - windows`
